WIP: Define VDAF for heavy hitters (#4)
cfrg/draft-irtf-cfrg-vdaf: draft-patton-cfrg-vdaf.md

* Output shares are secret shares of a vector of field elements, each
  corresponding to a counter for one of the candidate prefixes.

## Dependencies

[TODO Clarify which of these can be assumed and which need to be explicitly defined]

> ### Descriptions of finite fields

Reviewer: Add a newline between headings and the first line.

Author: Done.

We assume a structure for describing DPF output groups, named `ValueType`.
For verifiable heavy hitters, we need an IDPF that at each level has two field 
elements as outputs, and so the value type would be pairs of field elements.
The IDPF key generation and evaluation functions will allow one `ValueType` parameter
per bit in the IDPF domain, which allows us to use different types at the output

> layer, required for extractability. See [Section 4.3](https://eprint.iacr.org/2021/017.pdf) in [BBCGGI21] for details, and see [here](https://github.com/google/distributed_point_functions/blob/master/dpf/distributed_point_function.proto#L25) for an example implementation of `ValueType`. We use `ValueType` to refer to a description of a given group, and `Value` to refer to an actual element of the group.

Reviewer: * Replace this reference with `[BBCGGI21, Section 4.3]`
* When this gets compiled, the markdown reference won't appear in the text version. The reference should be added to the  list of informative references at the top of this file.

Author: I removed the second reference, since the value type implementation is not really important if we assume that the concrete scheme fixes the types for each level.

### Incremental Distributed Point Functions (IDPFs)

> An IDPF is defined over a domain of size `2**d`. The client can specify an index `alpha` and values `beta`, one for each level `l` in `[d]`. The key generation generates two DPF keys that individually hide `alpha` and `beta`. When locally evaluated at any point `x` in `2**l` at level `l`, the IDPF returns shares of `beta[l]`, if `x` is the `l`-bit prefix of `alpha`, and shares of zero otherwise.

Reviewer: ?
```suggestion
An IDPF is defined over a domain of size `2**d`. The client can specify an index `alpha` and values `beta`, one for each level `l` in `[d]`. The key generation generates two IDPF keys that individually hide `alpha` and `beta`. When locally evaluated at any point `x` in `2**l` at level `l`, the IDPF returns shares of `beta[l]`, if `x` is the `l`-bit prefix of `alpha`, and shares of zero otherwise.
```

If elements of `beta` are tuples, the additive sharing is done element-wise.

> - `idpf_gen(value_types: Vec[ValueType], alpha: int, beta: Vec[Value]) -> (Bytes, Bytes)`: Takes as input the value types describing the IDPF output domains, as well as the index and values for the IDPF. Returns two serialized DPF keys, one for each helper server.

Reviewer: ```suggestion
- `idpf_gen(value_types: Vec[ValueType], alpha: int, beta: Vec[Value]) -> (Bytes, Bytes)`: Takes as input the value types describing the IDPF output domains, as well as the index and values for the IDPF. Returns two serialized DPF keys, one for each aggregator.
```

> - `idpf_eval_next(state: State, dpf_key: Bytes, value_type: ValueType, x: int) -> (new_state: State, value: Value)`: Evaluates the given key at index `x`, at the next prefix length that hasn't beeen evaluated yet for `state`, returning a secret-shared value of the given type.

Reviewer: Should `x` be unsigned? How about `Unsigned` as the type hint, for consistency with the section above?

> ### Malicious sketching for IDPFS

Reviewer: ```suggestion
### Malicious sketching for IDPFs
```

  return (public_param, verify_param)
~~~

~~~
def vdaf_input(public_param, input):
  assert(log2(input) < public_param.n)
  beta = []
  correlation_shares_0 = [], correlation_shares_1 = []
  for l in range(public_param.n):
    # Construct values of the form (1, k), where k is a random field element.
    _, field = public_param.value_types[l]
    k = get_random_element(field)
    beta += [(1, k)]
    # Create secret shares of correlations to aid the servers' computation
    a, b, c = get_random_element(field), get_random_element(field), get_random_element(field)
    A = -2*a+k
    B = a*a + b - a * k + c
    correlation_shares_0 += [{
      a: get_random_element(field),
      b: get_random_element(field),
      c: get_random_element(field),
      A: get_random_element(field),
      B: get_random_element(field),
    }]
    correlation_shares_1 += [{
      a: a - correlation_shares_0[l]
      b: b - correlation_shares_0[l]
      c: c - correlation_shares_0[l]
      A: A - correlation_shares_0[l]
      B: B - correlation_shares_0[l]
    }]
  key_0, key_1 = idpf_gen(public_param.value_types, input, beta)
  return ((key_0, correlation_shares_0), (key_1, correlation_shares_1))
~~~

> [TODO: Not sure what to do with `nonce`]

Reviewer: The nonce should be used to generate a seed for the verification randomness vector `r`. For example, in prio3 we have
```python
def vdaf_start(k_query_init, _, nonce, r_input_share):  
  ...
  k_query_rand = get_key(k_query_init, byte(255) + nonce)
  ...
```

where `k_query_init` is the verification parameter.

Author: Done.

* Output shares are secret shares of a vector of field elements, each
  corresponding to a counter for one of the candidate prefixes.

## Dependencies

[TODO Clarify which of these can be assumed and which need to be explicitly
defined]

### Descriptions of finite fields

We assume a structure for describing IDPF output groups, named `ValueType`.
For verifiable heavy hitters, we need an IDPF that at each level has two field
elements as outputs, and so the value type would be pairs of field elements.
The IDPF key generation and evaluation functions will allow a different `ValueType`
parameter per bit in the IDPF domain, which allows us to use different types at
the output layer, required for extractability. See [BBCGGI21, Section 4.3] for

> details. We assume a function `get_value_type(n: Unsigned) -> ValueType` that

Reviewer: ```suggestion
details. An instantiation of this VDAF specifies a function `get_value_type(n: Unsigned) -> ValueType` that
```

returns the type for bit index `n`. We use `ValueType` to refer to a description

> of a given group, and`Value` to refer to an actual element of the group.

Reviewer: ```suggestion
of a given group, and `Value` to refer to an actual element of the group.
```

### Incremental Distributed Point Functions (IDPFs)

> An IDPF is defined over a domain of size `2**d`. The client can specify an

Reviewer: I know we're supposed to be doing everything in Python, but let's just write this as
```suggestion
An IDPF is defined over a domain of size `2^d`. The client can specify an
```

index `alpha` and values `beta`, one for each level `l` in `[d]`. The key
generation generates two IDPF keys that individually hide `alpha` and `beta`.

> When locally evaluated at any point `x` in `2**l` at level `l`, the IDPF
> returns shares of `beta[l]`, if `x` is the `l`-bit prefix of `alpha`, and

Reviewer: Would it be more accurate to say
```suggestion
When locally evaluated at any point `x` in `2**l` at level `l`, the IDPF
key returns a share of `beta[l]`, if `x` is the `l`-bit prefix of `alpha`, and
```

shares of zero otherwise. If elements of `beta` are tuples, the additive
sharing is done element-wise.

> - `idpf_gen(alpha: Unsigned, beta: Vec[Value]) -> (Bytes, Bytes)`:

Reviewer: I've been naming the output parameters. How about `-> (idpf_key1: Bytes, idpf_key2: Bytes)` or something?

Takes as input the index and values for the IDPF. Returns two serialized IDPF

> keys, one for each helper server.

Reviewer: ```suggestion
keys, one for each aggregator.
```

> - `idpf_eval_next(state: State, dpf_key: Bytes, x: Unsigned) `

Reviewer: ```suggestion
- `idpf_eval_next(state: State, idpf_key: Bytes, x: Unsigned) `
```

`-> (new_state: State, value: Value)`:
Evaluates the given key at index `x`, at the next prefix length that hasn't
been evaluated yet for `state`, returning a secret-shared value.

### Malicious sketching for IDPFs

~~~
def verify_sketch_round_1(dpf_output: Vec[Value], correlation_share, joint_randomness: Bytes, level: Unsigned):
  # Expand common randomness and compute verification message

>   r = expand_rand(joint_randomness, len(dpf_output), level)

Reviewer: Above I use "joint rand" to refer to randomness shared by the prover and verifier for PCPs. Can we call this something else, like "verify_rand"?

>   r = expand_rand(joint_randomness, len(dpf_output), level)

Reviewer: Here and elsewhere
```suggestion
  r = expand_rand(joint_randomness, len(idpf_output), level)
```

  r_squared = [v*v for v in r]
  ver_msg_1 = inner_product(

>     [value[0] for value in dpf_values,

Reviewer: Here and below
```suggestion
    [value[0] for value in dpf_output,
```

  for l in range(public_param.n):
    # Construct values of the form (1, k), where k is a random field element.
    _, field = get_value_type(l)
    k = get_random_element(field)
    beta += [(1, k)]
    # Create secret shares of correlations to aid the servers' computation
    a, b, c = get_random_element(field), get_random_element(field), \
      get_random_element(field)
    A = -2*a+k
    B = a*a + b - a * k + c
    correlation_shares_0 += [{
      a: get_random_element(field),
      b: get_random_element(field),
      c: get_random_element(field),
      A: get_random_element(field),
      B: get_random_element(field),
    }]
    correlation_shares_1 += [{
      a: a - correlation_shares_0[l]
      b: b - correlation_shares_0[l]
      c: c - correlation_shares_0[l]
      A: A - correlation_shares_0[l]
      B: B - correlation_shares_0[l]
    }]
  key_0, key_1 = idpf_gen(input, beta)
  return ((key_0, correlation_shares_0), (key_1, correlation_shares_1))
~~~

~~~
def vdaf_start(verify_param, public_param, nonce, input_share):
  key, correlation_share = input_share

  # Expand left and right children of the DPF tree.
  state_left, value_left = idpf_eval_next(None, key, 0)

>   state_right, value_right = idpf_eval_next(None, key, 1)

Reviewer: I'm having a hard time understanding what's going on here. It seeems like you want the IDPF api to be stateful, but the VDAF api is not. So how does the state of the IPDF get passed through to subsequent VDAF evaluations?

Unless I'm misunderstanding, I think the answer is that, at least as far as this spec is concerned, the IDPF api needs to be stateless.
